apps/explorer/lib/explorer/market/history/cataloger.ex

```diff
@@ -28,38 +28,67 @@ defmodule Explorer.Market.History.Cataloger do
 
   @typep milliseconds :: non_neg_integer()
 
+  @market_cap_failed_attempts 3
+
   @impl GenServer
   def init(:ok) do
-    send(self(), {:fetch_history, 365})
+    send(self(), {:fetch_price_history, 365})
 
     {:ok, %{}}
   end
 
   @impl GenServer
-  def handle_info({:fetch_history, day_count}, state) do
-    fetch_history(day_count)
+  def handle_info({:fetch_price_history, day_count}, state) do
+    fetch_price_history(day_count)
+
+    {:noreply, state}
+  end
+
+  @impl GenServer
+  def handle_info(:fetch_market_cap_history, state) do
+    fetch_market_cap_history()
 
     {:noreply, state}
   end
 
   @impl GenServer
   # Record fetch successful.
-  def handle_info({_ref, {_, _, {:ok, records}}}, state) do
-    Market.bulk_insert_history(records)
+  def handle_info({_ref, {:price_history, {_, _, {:ok, records}}}}, state) do
+    Process.send(self(), :fetch_market_cap_history, [])
+    state = state |> Map.put_new(:price_records, records)
 
-    # Schedule next check for history
-    fetch_after = config_or_default(:history_fetch_interval, :timer.minutes(60))
-    Process.send_after(self(), {:fetch_history, 1}, fetch_after)
+    {:noreply, state |> Map.put_new(:price_records, state)}
+  end
+
+  @impl GenServer
+  # Record fetch successful.
+  def handle_info({_ref, {:market_cap_history, {_, {:ok, market_cap_records}}}}, state) do
+    records = compile_records(state.price_records, market_cap_records)
+    market_cap_history(records, state)
+  end
+
+  @impl GenServer
+  # Record fetch successful.
+  def handle_info({:market_cap_history, {_, {:ok, nil}}}, state) do
```

Comment:
Seems like you missed `_ref` tuple:
```suggestion
  def handle_info({_ref, {:market_cap_history, {_, {:ok, nil}}}}, state) do
```
If yes, then this clause should be placed above this `handle_info({_ref, {:market_cap_history, {_, {:ok, market_cap_records}}}}, state)`

Reply:
nice catch, fixed